## Requirements
The product that you produce for this assignment is a notebook that describes your approach, as well as the code that implements the approach. Your notebook needs to include the following sections and information:

1. Introduction - A summary of your general approach
2. Data and statistics - Describe the data and statistics that you used, and how they were used.
3. Results of your approach - This section needs to include the top 5 players that your approach identifies. You also need to select one player from your top 5 and provide an analysis of his rating and monetary value.
4. Limitations - What difficulties do you see with implementing your strategy in the draft?

The draft will occur in class on March 20. Your assignment needs to be submitted before class on that day. 

To make grading easier, please name the notebook *Assignment4_LastName1_LastName2*, where *LastName1* and *LastName2* are the last names of the team members. Yes, you will be working in teams of two for this assignment.

## Introduction

Our approach will be to exploit the high points given for singles, doubles, triples, and homeruns. In particular, we will look for players who can get us these points. 

| Batting Stat | Points | 
|--------------|--------|
| H            | 5.6    |
| 2B           | 2.9    |
| 3B           | 5.7    |
| HR           | 9.4    |

<br>


In order to rank our fielding players, we will take inspiration from wOBA by using linear weights. Since we have limited funds to build our team, we are particularly interested in which players can give us the most amount of points for least amount of money. We will use the following multipliers (which correspond to the pointing system for the fantasy scoreing) for H, 2B, 3B, and HR and divide by salary. Our fomula will use multipliers and follow the formula below: 

| Event | Multiplier | 
|-------|--------|
| H     | 5.6    |
| 2B    | 2.9    |
| 3B    | 5.7    |
| HR    | 9.4    |


<br>

$$
\frac{(5.6 * H) + (2.9 * 2B) + (5.7 * 3B) + (9.4 * HR) }{salary}
$$
***
In order to rank our pitchers, we will take inspiration from FIP. FIP uses multipliers for different statistics which we will use in our formula. We will exploit the points below & calculate how many points per dollar each pitcher will bring us. See the formula below.

| Pitching Stat | Points |
|---------------|--------|
| IP            | 5      |
| K             | 2.0    |
| BB            | -3.0   |
| HBP           | -3.0   |
| HR            | -13    |
<br>
$$
\frac{(5.0 * IP) + (2.0 * K) + (-3.0 * (BB+HBP)) + (-13 * HR) }{salary}
$$

We will apply this formula for every player that has played in at least 100 games & sort them by their repected positions. 


## Data
### Pitching & Batting for 2018

We used the 2018 statcast hitter & 2016 Lahman pitcher data and Fan Graph salary projections for fantasy drafting data. From statcast we grabbed the statistics needed for our formula (H, 2B, 3B, HR). From Lahman, we grabbed the statistics needed for out formula (IP, K, BB, HBP, HR). From Fan Graphs, we grabbed data for positions of the players and their respective salaries. 

In [1]:
from pybaseball import statcast
from pybaseball import batting_stats_range
import matplotlib.pylab as plt
import pandas as pd
import numpy as np

pitching2018 = pd.read_csv('statcast_2018.csv')
batting2018 = pd.read_csv('statcast_batting_2018.csv')
hittersSalaries = pd.read_csv('batters_fan_graph.csv')
pitchersSalaries = pd.read_csv('pitchers_fan_graph.csv')

pd.set_option('display.max_columns', None)
batting2018 = batting2018.loc[batting2018['G']>=50] # more than 100 games played

### Salary Data

In [2]:
# convert 'salary' string column to floats by removing ($,())
hittersSalaries['Dollars'] = [x.strip('$') for x in hittersSalaries.Dollars]
hittersSalaries['Dollars'] = [x.strip('($') for x in hittersSalaries.Dollars]
hittersSalaries['Dollars'] = [x.strip(')') for x in hittersSalaries.Dollars]
hittersSalaries['Dollars'] = pd.to_numeric(hittersSalaries['Dollars'])
hittersSalaries

PlayerName          Team          POS    ADP   PA    mAVG  \
0         Mookie Betts       Red Sox           OF    1.9  666    $8.1   
1           Mike Trout        Angels           OF    1.1  595    $6.2   
2         Jose Ramirez       Indians           3B    3.9  657    $4.2   
3        J.D. Martinez       Red Sox        OF/DH    5.5  592    $6.3   
4        Nolan Arenado       Rockies           3B    7.2  653    $4.6   
5     Francisco Lindor       Indians           SS   15.2  652    $4.6   
6    Giancarlo Stanton       Yankees        OF/DH   21.5  602    $0.6   
7          Trea Turner     Nationals           SS    7.3  658    $4.9   
8     Christian Yelich       Brewers           OF    7.1  641    $6.7   
9     Ronald Acuna Jr.        Braves           OF    8.8  652    $3.2   
10       J.T. Realmuto      Phillies            C   47.3  536    $0.9   
11        Trevor Story       Rockies           SS   17.0  607    $1.4   
12    Charlie Blackmon       Rockies           OF   27.5  678    $5.0   
13        Gary Sanchez       Yankees            C   56.0  492  ($3.1)   
14       Manny Machado        Padres           SS   18.2  637    $3.2   
15       Anthony Rizzo          Cubs           1B   37.2  644    $3.4   
16        Bryce Harper      Phillies           OF   18.1  614  ($0.7)   
17        Alex Bregman        Astros        3B/SS   13.8  649    $3.1   
18         Javier Baez          Cubs     2B/3B/SS   17.0  627    $1.1   
19         Jose Altuve        Astros           2B   13.9  653    $8.3   
20     Freddie Freeman        Braves           1B   21.2  646    $4.5   
21         Nelson Cruz         Twins           DH   97.0  600    $3.5   
22   Andrew Benintendi       Red Sox           OF   29.2  655    $4.7   
23      Starling Marte       Pirates           OF   36.0  624    $3.7   
24           Juan Soto     Nationals           OF   35.2  602    $5.1   
25    Paul Goldschmidt     Cardinals           1B   18.8  648    $2.6   
26        Buster Posey        Giants            C  144.3  531    $3.9   
27         Kris Bryant          Cubs        3B/OF   32.2  629    $2.1   
28         Aaron Judge       Yankees           OF   15.3  619  ($2.6)   
29         Yasiel Puig          Reds           OF   71.6  575    $2.2   
..                 ...           ...          ...    ...  ...     ...   
470      Alex Blandino          Reds           2B  601.0   90  ($1.0)   
471       Carlos Gomez          Mets           OF  600.5   47  ($0.7)   
472   Brandon Phillips           NaN        2B/3B  999.0   71  ($0.1)   
473       Carlos Tocci       Rangers           OF  999.0   52  ($0.3)   
474    Ronald Torreyes         Twins           2B  999.0   71    $0.1   
475          A.J. Reed        Astros           1B  597.8   40  ($0.6)   
476       Mark Zagunis          Cubs           OF  999.0   45  ($0.5)   
477       Josh Fuentes       Rockies           1B  999.0   28    $0.2   
478      Tyrone Taylor       Brewers           OF  999.0   40  ($0.5)   
479       LaMonte Wade         Twins           OF  999.0   33  ($0.1)   
480      Joey Curletta      Mariners        1B/OF  999.0   39  ($0.5)   
481        Cole Tucker       Pirates           SS  600.8   34  ($0.3)   
482   Sergio Alcantara        Tigers           SS  999.0   46  ($0.4)   
483       Willi Castro        Tigers           SS  999.0   33  ($0.2)   
484     Breyvic Valera        Giants  2B/3B/SS/OF  999.0   32    $0.0   
485        Alen Hanson        Giants        2B/OF  578.0   53  ($0.3)   
486       Michael Reed         Twins           OF  999.0   27  ($0.1)   
487       Ryan Cordell     White Sox           OF  999.0   33  ($0.4)   
488        Matt Thaiss        Angels           1B  598.1   33  ($0.2)   
489      Bobby Bradley       Indians           1B  601.0   26  ($0.3)   
490        Brad Miller       Dodgers        1B/2B  999.0   61  ($0.8)   
491       Dylan Cozens      Phillies           OF  999.0   27  ($0.5)   
492        Jose Pirela        Padres        2B/OF  600.1   49  ($0.1) 

In [3]:
# convert 'salary' string column to floats by removing ($,())
pitchersSalaries['Dollars'] = [x.strip('$') for x in pitchersSalaries.Dollars]
pitchersSalaries['Dollars'] = [x.strip('($') for x in pitchersSalaries.Dollars]
pitchersSalaries['Dollars'] = [x.strip(')') for x in pitchersSalaries.Dollars]
pitchersSalaries['Dollars'] = pd.to_numeric(pitchersSalaries['Dollars'])
pitchersSalaries.head()

PlayerName       Team POS   ADP   IP    mW     mSV   mERA  mWHIP  \
0        Chris Sale    Red Sox  SP  13.1  184  $5.9  ($2.0)  $11.1  $14.1   
1      Max Scherzer  Nationals  SP   4.7  208  $5.8  ($2.0)   $7.7  $11.6   
2      Jacob deGrom       Mets  SP  10.6  208  $4.9  ($2.0)  $10.4  $10.1   
3  Justin Verlander     Astros  SP  20.9  202  $5.6  ($2.0)   $4.4   $9.4   
4      Corey Kluber    Indians  SP  24.4  209  $5.2  ($2.0)   $3.3   $6.6   

    mSO    PTS   aPOS  Dollars  
0  $7.0  $36.0  $10.3     47.3  
1  $8.3  $31.3  $10.3     42.6  
2  $7.1  $30.5  $10.3     41.8  
3  $7.0  $24.4  $10.3     35.7  
4  $5.1  $18.2  $10.3     29.5

# Applying the Formulas
***

## Fielding

In [4]:
#batting2018['fantasy'] = ((batting2018['H']*0.88)+(batting2018['2B']*1.247)+(batting2018['3B']*1.578)+(batting2018['HR']*2.031))/hittersSalaries['Dollars']
result = pd.merge(batting2018, hittersSalaries, how='inner', left_on='Name', right_on='PlayerName')
result['fantasy'] = ((result['H']*5.6)+(result['2B']*2.9)+(result['3B']*5.7)+(result['HR']*9.4))/result['Dollars']

print(batting2018.shape)
#result = result.dropna()
result = result[['Name', 'POS', 'fantasy', 'Dollars']].copy()
result.sort_values(by=['fantasy'], ascending=False)

(398, 28)


Name          POS      fantasy  Dollars
290           Luke Voit           1B          inf      0.0
152     Kevin Kiermaier           OF  7896.000000      0.1
171       Nick Markakis           OF  5534.000000      0.2
119      Billy Hamilton           OF  4780.000000      0.2
46         Jason Castro            C  3197.000000      0.2
72      Corey Dickerson           OF  2603.800000      0.5
15        Austin Barnes            C  2414.500000      0.2
234      Hunter Renfroe           OF  2248.500000      0.4
35         Byron Buxton           OF  1760.800000      0.5
252     Domingo Santana           OF  1479.625000      0.8
217        Kevin Pillar           OF  1379.250000      0.8
129     Cesar Hernandez           2B  1292.750000      0.8
117        Yuli Gurriel        1B/3B  1184.400000      1.0
276        Chris Taylor        SS/OF  1153.300000      1.0
189     Kendrys Morales           DH  1114.100000      1.0
6         Yonder Alonso           1B   999.000000      1.0
164     Jonathan Lucroy            C   952.000000      0.8
197       Brandon Nimmo           OF   855.500000      0.4
47       Starlin Castro           2B   733.153846      1.3
105           Yan Gomes            C   686.111111      0.9
108     Marwin Gonzalez  1B/2B/SS/OF   679.750000      1.6
36     Asdrubal Cabrera     2B/3B/SS   649.866667      1.5
51    Robinson Chirinos            C   642.666667      0.9
154        Jason Kipnis           2B   622.100000      1.0
131      Odubel Herrera           OF   610.722222      1.8
76         Brian Dozier           2B   593.304348      2.3
57          Zack Cozart           3B   590.611111      1.8
10         Elvis Andrus           SS   586.666667      2.4
288   Christian Vazquez            C   579.090909      1.1
300      Ryan Zimmerman           1B   529.800000      2.5
..                  ...          ...          ...      ...
299      Bradley Zimmer           OF    22.888412     23.3
241        Miguel Rojas     1B/3B/SS    22.750000     22.4
220         Jose Pirela        2B/OF    22.731148     30.5
70         Aledmys Diaz        3B/SS    22.247899     23.8
126  Adeiny Hechavarria           SS    21.961404     28.5
138        Rhys Hoskins           OF    21.795000     20.0
185      Devin Mesoraco            C    21.646018     11.3
170      Jake Marisnick           OF    21.068376     23.4
279     Ronald Torreyes           2B    19.885906     29.8
88       Logan Forsythe           2B    19.286689     29.3
275        Raimel Tapia        OF/DH    18.540230     17.4
298         Chris Young           OF    18.427350     23.4
240    Daniel Robertson        2B/SS    18.071038     18.3
186         Brad Miller        1B/2B    17.225166     30.2
249      Pablo Sandoval        1B/3B    16.197026     26.9
287         Pat Valaika  1B/2B/3B/SS    15.927481     26.2
201          Jose Osuna     1B/3B/OF    15.748062     25.8
147        JaCoby Jones           OF    14.508197     12.2
99          Greg Garcia           2B    14.312950     27.8
238         T.J. Rivera        2B/3B    13.761006     31.8
42           Mark Canha           OF    12.751092     22.9
121         Alen Hanson        2B/OF    11.614618     30.1
2        Ehire Adrianza        3B/SS    11.127341     26.7
156     Tommy La Stella           3B    10.959677     24.8
247      Tyler Saladino           SS    10.376344     27.9
137          Brock Holt        2B/SS    10.187135     17.1
213       Jace Peterson        3B/OF     9.800699     28.6
245            Rio Ruiz           3B     8.580000     25.0
85         Derek Fisher           OF     8.466192     28.1
191          Max Moroff        2B/SS     7.138889     25.2

[303 rows x 4 columns]

## Catchers

In [5]:
resultC = result.loc[result['POS']=='C'].copy().sort_values(by='fantasy', ascending=False)
resultC

Name POS      fantasy  Dollars
46         Jason Castro   C  3197.000000      0.2
15        Austin Barnes   C  2414.500000      0.2
164     Jonathan Lucroy   C   952.000000      0.8
105           Yan Gomes   C   686.111111      0.9
51    Robinson Chirinos   C   642.666667      0.9
288   Christian Vazquez   C   579.090909      1.1
127       Austin Hedges   C   525.615385      1.3
273         Kurt Suzuki   C   408.187500      1.6
167    Martin Maldonado   C   344.952381      2.1
215        Josh Phegley   C   285.125000      0.8
180        Brian McCann   C   261.576923      2.6
87        Tyler Flowers   C   219.200000      3.0
61      Travis d'Arnaud   C   208.242424      3.3
13           Alex Avila   C   186.529412      3.4
45   Welington Castillo   C   180.357143      4.2
141      Chris Iannetta   C   167.027778      3.6
196        Omar Narvaez   C   125.657143      3.5
262         Kevan Smith   C   113.847826      4.6
48   Francisco Cervelli   C   110.880952      4.2
293        Matt Wieters   C   110.322581      6.2
302         Mike Zunino   C   103.743902      8.2
16      Tucker Barnhart   C    99.549296      7.1
181        James McCann   C    98.911765      6.8
218          Manny Pina   C    80.573171      8.2
212       Roberto Perez   C    77.021277      4.7
177         Jeff Mathis   C    65.860465      4.3
295        Tony Wolters   C    64.788462      5.2
160          Sandy Leon   C    63.098592      7.1
187       Yadier Molina   C    57.649718     17.7
140        Nick Hundley   C    54.878788      9.9
55    Willson Contreras   C    54.590909     15.4
243       Austin Romine   C    54.196721      6.1
71           Elias Diaz   C    52.814815      5.4
225        Buster Posey   C    51.224299     21.4
112     Yasmani Grandal   C    47.335106     18.8
132      Chris Herrmann   C    45.853333      7.5
237         Rene Rivera   C    45.063158      9.5
248        Gary Sanchez   C    44.072000     25.0
231       J.T. Realmuto   C    42.906977     25.8
230        Wilson Ramos   C    32.806202     12.9
155        Andrew Knapp   C    30.969388      9.8
185      Devin Mesoraco   C    21.646018     11.3

## 1B

In [6]:
result1B = result.loc[result['POS']=='1B'].copy().sort_values(by='fantasy', ascending=False)
result1B

Name POS     fantasy  Dollars
290         Luke Voit  1B         inf      0.0
6       Yonder Alonso  1B  999.000000      1.0
300    Ryan Zimmerman  1B  529.800000      2.5
21       Brandon Belt  1B  259.933333      3.0
264      Justin Smoak  1B  243.886792      5.3
190    Mitch Moreland  1B  164.819672      6.1
63        Chris Davis  1B  135.870968      6.2
139       Eric Hosmer  1B  130.166667     10.8
19          Josh Bell  1B  119.789474      9.5
86      Wilmer Flores  1B  114.656250      6.4
251    Carlos Santana  1B  111.371429     10.5
125        Ryon Healy  1B  103.713043     11.5
291        Joey Votto  1B   98.979452     14.6
3       Jesus Aguilar  1B   92.492537      6.7
0          Jose Abreu  1B   83.469945     18.3
27        Justin Bour  1B   80.142857     11.2
38     Miguel Cabrera  1B   79.036364     11.0
104  Paul Goldschmidt  1B   63.472477     21.8
1          Matt Adams  1B   56.810219     13.7
239     Anthony Rizzo  1B   53.736402     23.9
93    Freddie Freeman  1B   49.221739     23.0
200        Matt Olson  1B   36.919708     13.7
228    Hanley Ramirez  1B   33.489510     28.6

## 2B

In [7]:
result2B = result.loc[result['POS']=='2B'].copy().sort_values(by='fantasy', ascending=False)
result2B

Name POS      fantasy  Dollars
129  Cesar Hernandez  2B  1292.750000      0.8
47    Starlin Castro  2B   733.153846      1.3
154     Jason Kipnis  2B   622.100000      1.0
76      Brian Dozier  2B   593.304348      2.3
254  Jonathan Schoop  2B   430.757576      3.3
159      DJ LeMahieu  2B   343.777778      3.6
163       Jed Lowrie  2B   238.775510      4.9
103  Scooter Gennett  2B   233.255319      4.7
204        Joe Panik  2B   168.344262      6.1
188     Yoan Moncada  2B   153.083333      2.4
43     Robinson Cano  2B   125.404040      9.9
124    Josh Harrison  2B   108.488889      9.0
153      Ian Kinsler  2B    99.456311     10.3
208   Dustin Pedroia  2B    85.576087      9.2
296      Kolten Wong  2B    85.195122      8.2
199     Rougned Odor  2B    79.278195     13.3
194    Daniel Murphy  2B    68.812500     19.2
289  Jonathan Villar  2B    65.231481     10.8
8        Jose Altuve  2B    64.823276     23.2
5       Ozzie Albies  2B    36.656000     12.5
150   Howie Kendrick  2B    28.932489     23.7
280     Devon Travis  2B    25.915493     14.2
279  Ronald Torreyes  2B    19.885906     29.8
88    Logan Forsythe  2B    19.286689     29.3
99       Greg Garcia  2B    14.312950     27.8

## 3B

In [8]:
result3B = result.loc[result['POS']=='3B'].copy().sort_values(by='fantasy', ascending=False)
result3B

Name POS     fantasy  Dollars
57       Zack Cozart  3B  590.611111      1.8
162    Evan Longoria  3B  342.800000      3.5
257      Kyle Seager  3B  313.945946      3.7
250      Miguel Sano  3B  295.406250      3.2
158        Jake Lamb  3B  291.435897      3.9
90     Maikel Franco  3B  131.825000      8.0
75    Josh Donaldson  3B  101.867347      9.8
193   Mike Moustakas  3B   92.202899     13.8
272   Eugenio Suarez  3B   80.195652     13.8
284    Justin Turner  3B   78.964539     14.1
233   Anthony Rendon  3B   76.987342     15.8
118      Jedd Gyorko  3B   67.902985     13.4
92      Todd Frazier  3B   58.680000     15.0
41     Johan Camargo  3B   54.010526      9.5
12     Nolan Arenado  3B   51.302632     30.4
50      Matt Chapman  3B   48.401639     12.2
69     Rafael Devers  3B   47.784314     10.2
229     Jose Ramirez  3B   46.919255     32.2
156  Tommy La Stella  3B   10.959677     24.8
245         Rio Ruiz  3B    8.580000     25.0

## SS

In [9]:
resultSS = result.loc[result['POS']=='SS'].copy().sort_values(by='fantasy', ascending=False)
resultSS

Name POS     fantasy  Dollars
10         Elvis Andrus  SS  586.666667      2.4
58     Brandon Crawford  SS  484.750000      2.0
261   Andrelton Simmons  SS  225.307692      5.2
65          Paul DeJong  SS  223.295455      4.4
11        Orlando Arcia  SS  209.723404      4.7
96        Freddy Galvis  SS  142.740260      7.7
9          Tim Anderson  SS  141.487179      7.8
274      Dansby Swanson  SS  108.070423      7.1
259       Marcus Semien  SS   98.562500      6.4
56        Carlos Correa  SS   95.327273     11.0
114      Didi Gregorius  SS   85.408759     13.7
223       Jorge Polanco  SS   80.547826     11.5
256        Corey Seager  SS   79.526667     15.0
210         Jose Peraza  SS   73.532110     10.9
258         Jean Segura  SS   72.066667     15.0
183        Jordy Mercer  SS   63.100000     15.0
25      Xander Bogaerts  SS   62.193182     17.6
165       Manny Machado  SS   53.619433     24.7
4            Nick Ahmed  SS   51.693548      6.2
161    Francisco Lindor  SS   48.099010     30.3
246     Addison Russell  SS   41.584906     15.9
271        Trevor Story  SS   39.665354     25.4
142       Jose Iglesias  SS   32.228346     25.4
285         Trea Turner  SS   29.840830     28.9
283     Troy Tulowitzki  SS   25.951807     16.6
126  Adeiny Hechavarria  SS   21.961404     28.5
247      Tyler Saladino  SS   10.376344     27.9

## OF

In [10]:
resultOF = result.loc[result['POS']=='OF'].copy().sort_values(by='fantasy', ascending=False)
resultOF

Name POS      fantasy  Dollars
152     Kevin Kiermaier  OF  7896.000000      0.1
171       Nick Markakis  OF  5534.000000      0.2
119      Billy Hamilton  OF  4780.000000      0.2
72      Corey Dickerson  OF  2603.800000      0.5
234      Hunter Renfroe  OF  2248.500000      0.4
35         Byron Buxton  OF  1760.800000      0.5
252     Domingo Santana  OF  1479.625000      0.8
217        Kevin Pillar  OF  1379.250000      0.8
197       Brandon Nimmo  OF   855.500000      0.4
131      Odubel Herrera  OF   610.722222      1.8
28   Jackie Bradley Jr.  OF   496.000000      1.8
40         Kole Calhoun  OF   369.678571      2.8
169       Manuel Margot  OF   344.851852      2.7
232        Josh Reddick  OF   338.625000      3.2
110         Alex Gordon  OF   244.275862      2.9
107     Carlos Gonzalez  OF   208.863636      4.4
143      Ender Inciarte  OF   180.513514      7.4
266    Steven Souza Jr.  OF   178.694915      5.9
207        Joc Pederson  OF   156.838710      3.1
133       Jason Heyward  OF   150.358491      5.3
253      Scott Schebler  OF   148.772727      6.6
151          Max Kepler  OF   143.705882      6.8
101       Brett Gardner  OF   133.500000      8.8
286        Justin Upton  OF   124.552381     10.5
224        A.J. Pollock  OF   119.240000      7.5
54     Michael Conforto  OF   113.911392      7.9
222     Gregory Polanco  OF   111.838710      6.2
255      Kyle Schwarber  OF    94.602273      8.8
214          Tommy Pham  OF    90.775862     11.6
134         Aaron Hicks  OF    90.309859      7.1
..                  ...  ..          ...      ...
166       Mikie Mahtook  OF    43.082840     16.9
263        Mallex Smith  OF    41.777778     10.8
97            Ben Gamel  OF    38.684825     25.7
7         Aaron Altherr  OF    34.379592     24.5
23         Mookie Betts  OF    32.911392     39.5
52    Lonnie Chisenhall  OF    29.978142     18.3
26      Jorge Bonifacio  OF    29.372093     25.8
281          Mike Trout  OF    29.342318     37.1
49      Yoenis Cespedes  OF    28.900415     24.1
100        Leury Garcia  OF    28.242857     21.0
145             Jon Jay  OF    28.161417     25.4
294       Nick Williams  OF    27.317269     24.9
236        Joey Rickard  OF    26.950311     16.1
109       Brian Goodwin  OF    25.913876     20.9
173      Starling Marte  OF    25.717489     22.3
106        Carlos Gomez  OF    25.715254     29.5
178      Cameron Maybin  OF    25.669231     26.0
144      Austin Jackson  OF    25.661290     24.8
79         Jarrod Dyson  OF    25.170940     23.4
128   Guillermo Heredia  OF    24.536398     26.1
80      Jacoby Ellsbury  OF    24.035714     28.0
157        Juan Lagares  OF    23.216931     18.9
82           Adam Engel  OF    23.216867     16.6
299      Bradley Zimmer  OF    22.888412     23.3
138        Rhys Hoskins  OF    21.795000     20.0
170      Jake Marisnick  OF    21.068376     23.4
298         Chris Young  OF    18.427350     23.4
147        JaCoby Jones  OF    14.508197     12.2
42           Mark Canha  OF    12.751092     22.9
85         Derek Fisher  OF     8.466192     28.1

[90 rows x 4 columns]

# The Line-Up
- C  - Jason Castro
- 1B - Yonder Alonso
- 2B - Cesar Hernandez
- 3B - Zack Cozart
- SS - Elvis Andrus	
- OF - Kevin Kiermaier, Nick Markakis, Billy Hamilton
***




# Pitchers

In [11]:
from pybaseball.lahman import *
import pandas as pd

# data
download_lahman()
# only for Eric's machine
#pitching = pd.read_csv("C:/Users/Eric/Documents/CU/CU/sabermetrics/baseballdatabank-2017.1/core/Pitching.csv")

pitching = pitching()
pitching2018_L = pitching.loc[(pitching['yearID']==2016)].copy()
pitching2018_L

playerID  yearID  stint teamID lgID   W   L   G  GS  CG  SHO  SV  \
44139   abadfe01    2016      1    MIN   AL   1   4  39   0   0    0   1   
44140   abadfe01    2016      2    BOS   AL   0   2  18   0   0    0   0   
44141  achteaj01    2016      1    LAA   AL   1   0  27   0   0    0   0   
44142  adamsau01    2016      1    CLE   AL   0   0  19   0   0    0   0   
44143  adlemti01    2016      1    CIN   NL   4   4  13  13   0    0   0   
44144  alberan01    2016      1    MIN   AL   0   0   6   2   0    0   0   
44145  alberma01    2016      1    CHA   AL   2   6  58   1   0    0   0   
44146  albural01    2016      1    LAA   AL   0   0   2   0   0    0   0   
44147  alcanra01    2016      1    OAK   AL   1   3   5   5   0    0   0   
44148  alexasc01    2016      1    KCA   AL   0   0  17   0   0    0   0   
44149  allenco01    2016      1    CLE   AL   3   5  67   0   0    0  32   
44150  altavda01    2016      1    SEA   AL   0   0  15   0   0    0   0   
44151  alvarda01    2016      1    ATL   NL   3   1  16   0   0    0   0   
44152  alvarda01    2016      2    TEX   AL   0   0  10   0   0    0   0   
44153  alvarjo02    2016      1    LAA   AL   1   3  64   0   0    0   0   
44154  amarial01    2016      1    SDN   NL   0   0   1   0   0    0   0   
44155  anderbr04    2016      1    LAN   NL   1   2   4   3   0    0   0   
44156  anderch01    2016      1    MIL   NL   9  11  31  30   0    0   0   
44157  anderco01    2016      1    CLE   AL   2   5  19   9   0    0   0   
44158  anderty01    2016      1    COL   NL   5   6  19  19   0    0   0   
44159  andrima01    2016      1    TBA   AL   8   8  29  19   1    1   1   
44160  antoldu01    2016      1    TOR   AL   0   0   1   0   0    0   0   
44161  aquinja01    2016      1    BAL   AL   0   0   3   0   0    0   0   
44162  araujel01    2016      1    PHI   NL   2   1  32   0   0    0   0   
44163  archech01    2016      1    TBA   AL   9  19  33  33   0    0   0   
44164  armstsh01    2016      1    CLE   AL   0   0  10   0   0    0   0   
44165    arojo01    2016      1    SEA   AL   0   0   1   0   0    0   0   
44166  arrieja01    2016      1    CHN   NL  18   8  31  31   1    1   0   
44167  asheral01    2016      1    PHI   NL   2   1   5   5   0    0   0   
44168  avilalu01    2016      1    LAN   NL   3   0  27   0   0    0   0   
...          ...     ...    ...    ...  ...  ..  ..  ..  ..  ..  ...  ..   
44933  wilheto01    2016      1    TEX   AL   2   3  21   0   0    0   0   
44934  wilheto01    2016      2    SEA   AL   0   1  29   0   0    0   1   
44935  willije02    2016      1    SLN   NL   0   0  11   0   0    0   0   
44936  willitr01    2016      1    PIT   NL   1   1   7   1   0    0   0   
44937  wilsoal01    2016      1    DET   AL   4   0  62   0   0    0   0   
44938  wilsoju10    2016      1    DET   AL   4   5  66   0   0    0   1   
44939  wilsoty01    2016      1    BAL   AL   4   6  24  13   0    0   0   
44940  wimmeal01    2016      1    MIN   AL   1   3  16   0   0    0   0   
44941  winklda01    2016      1    ATL   NL   0   0   3   0   0    0   0   
44942  wislema01    2016      1    ATL   NL   7  13  27  26   0    0   1   
44943  withrch01    2016      1    ATL   NL   3   0  46   0   0    0   0   
44944  wittgni01    2016      1    MIA   NL   4   3  48   0   0    0   0   
44945   woodal02    2016      1    LAN   NL   1   4  14  10   0    0   0   
44946   woodbl01    2016      1    CIN   NL   6   5  70   0   0    0   1   
44947   woodtr01    2016      1    CHN   NL   4   0  77   0   0    0   0   
44948  worleva01    2016      1    BAL   AL   2   2  35   4   0    0   1   
44949  wrighda04    2016      1    CIN   NL   0   2   4   2   0    0   0   
44950  wrighda04    2016      2    LAA   AL   1   3   5   5   0    0   0   
44951  wrighmi01    2016      1    BAL   AL   3   4  18  12   0    0   0   
44952  wrighst01    2016      1    BOS   AL  13   6  24  24   4    1   0   
44953  yateski01    2016      1    NYA   AL   2   1  41 

In [12]:
pitching2018_L['IP'] = pitching2018_L['IPouts'] / 3
#master = pd.read_csv("C:/Users/Eric/Documents/CU/CU/sabermetrics/baseballdatabank-2017.1/core/Master.csv")
master = master()

In [13]:
pitching2018_L['fantasy'] = ((5.0*pitching2018_L['IP'])+(2.0*pitching2018_L['SO'])-(3.0*(pitching2018_L['BB']+pitching2018_L['HBP'])-(13.0*pitching2018_L['HR'])))


resultNames = pd.merge(pitching2018_L[['playerID', 'fantasy']], 
                      master[['playerID', 'nameFirst', 'nameLast']],
                      on='playerID', how='outer')
resultNames['Name'] = resultNames['nameFirst']+" "+resultNames['nameLast']


resultP = pd.merge(resultNames, pitchersSalaries, how='inner', left_on='Name', right_on='PlayerName')

resultP = resultP.dropna()
resultP = resultP[['Name', 'POS', 'fantasy', 'Dollars']].copy()
resultP.sort_values(by=['fantasy'], ascending=False)

Name POS      fantasy  Dollars
284       Max Scherzer  SP  1926.666667     42.6
325   Justin Verlander  SP  1841.333333     35.7
246        David Price  SP  1825.000000     11.5
36   Madison Bumgarner  SP  1787.333333     10.0
277         Chris Sale  SP  1764.333333     47.3
244      Rick Porcello  SP  1657.000000      6.6
9         Chris Archer  SP  1652.666667     16.7
165       Corey Kluber  SP  1623.000000     29.5
160        Ian Kennedy  SP  1538.333333     11.8
247      Jose Quintana  SP  1526.000000      9.1
172         Jon Lester  SP  1506.333333      0.4
278    Jeff Samardzija  SP  1497.666667      0.4
314    Masahiro Tanaka  SP  1497.333333     10.9
79         Danny Duffy  SP  1478.333333      2.1
297         Drew Smyly  SP  1473.666667      6.6
224      Jake Odorizzi  SP  1473.333333     10.2
241     Michael Pineda  SP  1466.333333      3.8
129        Cole Hamels  SP  1460.333333      7.3
106      Kevin Gausman  SP  1454.333333      2.8
308     Marcus Stroman  SP  1451.000000      2.5
303        Dan Straily  SP  1431.666667     11.8
188      Collin McHugh  SP  1425.333333      1.7
134  Jeremy Hellickson  SP  1412.000000      9.0
316      Julio Teheran  SP  1410.000000     11.3
329    Adam Wainwright  SP  1409.333333      1.2
251         Robbie Ray  SP  1388.666667     14.8
312   Noah Syndergaard  SP  1362.333333     19.3
257       Tanner Roark  SP  1357.000000      5.3
136     Kyle Hendricks  SP  1329.000000      6.8
11        Jake Arrieta  SP  1328.666667      1.0
..                 ...  ..          ...      ...
323       Jason Vargas  SP    86.000000      3.7
99       Carson Fulmer  SP    77.333333     17.8
265           Chaz Roe  RP    75.333333      5.9
235         Felix Pena  SP    75.000000      6.7
169       Wade LeBlanc  SP    74.000000      8.4
133      Andrew Heaney  SP    70.000000     11.4
170       Jose Leclerc  RP    66.000000      1.8
10     Shawn Armstrong  RP    65.333333      6.0
146       Tommy Hunter  RP    64.666667      5.3
305     Chris Stratton  SP    60.000000      3.6
152    Jeremy Jeffress  RP    59.666667      3.4
255    Clayton Richard  SP    57.000000     12.8
85        Roenis Elias  RP    55.333333      9.1
186     Kevin McCarthy  RP    53.666667     11.4
142    Jonathan Holder  RP    52.666667      8.0
248       Neil Ramirez  RP    47.333333      6.6
233        James Pazos  RP    45.666667      7.8
201        Juan Minaya  RP    42.666667      9.6
249       Neil Ramirez  RP    40.333333      6.6
274       Hyun-Jin Ryu  SP    38.333333      9.4
310      Daniel Stumpf  RP    36.000000      9.4
272         Tyson Ross  SP    27.666667     17.0
231      David Paulino  RP    27.000000      6.0
119         Matt Grace  RP    23.000000      8.2
179        Matt Magill  RP    21.666667     11.1
192    Yohander Mendez  SP     9.000000     18.3
189    Adalberto Mejia  RP     8.666667      9.1
225        Tyler Olson  RP     7.333333      6.6
128    Tayron Guerrero  RP     7.000000      7.3
229       Blake Parker  RP     2.000000      1.1

[306 rows x 4 columns]

## SP

In [14]:
resultSP = resultP.loc[resultP['POS']=='SP']
resultSP.sort_values(by=['fantasy'], ascending=False)

Name POS      fantasy  Dollars
284       Max Scherzer  SP  1926.666667     42.6
325   Justin Verlander  SP  1841.333333     35.7
246        David Price  SP  1825.000000     11.5
36   Madison Bumgarner  SP  1787.333333     10.0
277         Chris Sale  SP  1764.333333     47.3
244      Rick Porcello  SP  1657.000000      6.6
9         Chris Archer  SP  1652.666667     16.7
165       Corey Kluber  SP  1623.000000     29.5
160        Ian Kennedy  SP  1538.333333     11.8
247      Jose Quintana  SP  1526.000000      9.1
172         Jon Lester  SP  1506.333333      0.4
278    Jeff Samardzija  SP  1497.666667      0.4
314    Masahiro Tanaka  SP  1497.333333     10.9
79         Danny Duffy  SP  1478.333333      2.1
297         Drew Smyly  SP  1473.666667      6.6
224      Jake Odorizzi  SP  1473.333333     10.2
241     Michael Pineda  SP  1466.333333      3.8
129        Cole Hamels  SP  1460.333333      7.3
106      Kevin Gausman  SP  1454.333333      2.8
308     Marcus Stroman  SP  1451.000000      2.5
303        Dan Straily  SP  1431.666667     11.8
188      Collin McHugh  SP  1425.333333      1.7
134  Jeremy Hellickson  SP  1412.000000      9.0
316      Julio Teheran  SP  1410.000000     11.3
329    Adam Wainwright  SP  1409.333333      1.2
251         Robbie Ray  SP  1388.666667     14.8
312   Noah Syndergaard  SP  1362.333333     19.3
257       Tanner Roark  SP  1357.000000      5.3
136     Kyle Hendricks  SP  1329.000000      6.8
11        Jake Arrieta  SP  1328.666667      1.0
..                 ...  ..          ...      ...
57      Mike Clevinger  SP   382.000000      8.0
291       Tyler Skaggs  SP   338.333333      6.4
41      Andrew Cashner  SP   338.333333     30.9
334        Luke Weaver  SP   320.666667      0.0
256    Clayton Richard  SP   316.333333     12.8
173     Reynaldo Lopez  SP   290.000000     14.3
254         Alex Reyes  SP   278.000000      5.0
141          Rich Hill  SP   260.666667     16.9
64       Jharel Cotton  SP   232.666667      5.8
45      Jhoulys Chacin  SP   215.333333     11.2
16       Chris Bassitt  SP   209.000000      7.1
58           Alex Cobb  SP   186.000000     15.8
110      Lucas Giolito  SP   180.666667     21.1
70        Jose De Leon  SP   150.000000      4.9
112      Tyler Glasnow  SP   142.666667      4.1
87        Robbie Erlin  SP   134.333333      2.4
180     German Marquez  SP   132.333333     14.3
335    Trevor Williams  SP   122.333333     11.7
208     Charlie Morton  SP   113.666667     14.2
5       Brett Anderson  SP   106.666667      6.1
323       Jason Vargas  SP    86.000000      3.7
99       Carson Fulmer  SP    77.333333     17.8
235         Felix Pena  SP    75.000000      6.7
169       Wade LeBlanc  SP    74.000000      8.4
133      Andrew Heaney  SP    70.000000     11.4
305     Chris Stratton  SP    60.000000      3.6
255    Clayton Richard  SP    57.000000     12.8
274       Hyun-Jin Ryu  SP    38.333333      9.4
272         Tyson Ross  SP    27.666667     17.0
192    Yohander Mendez  SP     9.000000     18.3

[138 rows x 4 columns]

## RP

In [15]:
resultRP = resultP.loc[resultP['POS']=='RP']
resultRP.sort_values(by=['fantasy'], ascending=False)

Name POS      fantasy  Dollars
24       Chad Bettis  RP  1294.000000     10.9
276    Danny Salazar  RP  1021.666667      0.1
340    Steven Wright  RP   998.333333      9.7
31    Archie Bradley  RP   989.333333      1.0
206      Adam Morgan  RP   956.666667      6.9
236     Wily Peralta  RP   933.333333      7.9
215     Juan Nicasio  RP   905.000000      5.6
62       Adam Conley  RP   864.666667      7.9
127    Junior Guerra  RP   800.333333      6.0
73    Chris Devenski  RP   732.666667      4.7
130        Brad Hand  RP   661.666667      9.2
213     Hector Neris  RP   623.666667      4.6
240     David Phelps  RP   619.333333      8.0
12        Pedro Baez  RP   607.000000      5.9
226    Roberto Osuna  RP   600.000000      8.3
29        Brad Brach  RP   595.000000      7.5
23   Dellin Betances  RP   595.000000      7.6
252     Addison Reed  RP   583.333333     12.3
91     Michael Feliz  RP   579.000000      7.7
139   Kelvin Herrera  RP   565.000000      5.1
137  David Hernandez  RP   564.333333      7.3
150    Kenley Jansen  RP   564.333333     12.3
147  Raisel Iglesias  RP   555.666667      6.2
109        Ken Giles  RP   555.333333      5.0
153       Nate Jones  RP   550.333333      3.0
273      Chris Rusin  RP   546.666667      9.8
245     Ryan Pressly  RP   539.666667      0.9
259    Hansel Robles  RP   538.333333      8.9
2         Cody Allen  RP   537.000000      4.8
156     Shawn Kelley  RP   534.000000     10.5
..               ...  ..          ...      ...
81         Zach Duke  RP   123.666667      8.0
1    Scott Alexander  RP   119.000000      4.5
42     Miguel Castro  RP   118.333333     15.3
296       Will Smith  RP   116.666667      5.7
207   Brandon Morrow  RP   113.000000      5.1
317      Ryan Tepera  RP   107.666667      5.2
78    Brian Duensing  RP   103.666667     11.8
103      Luis Garcia  RP   103.666667      7.6
230     Blake Parker  RP    94.666667      1.1
96       Dylan Floro  RP    88.000000      7.0
265         Chaz Roe  RP    75.333333      5.9
170     Jose Leclerc  RP    66.000000      1.8
10   Shawn Armstrong  RP    65.333333      6.0
146     Tommy Hunter  RP    64.666667      5.3
152  Jeremy Jeffress  RP    59.666667      3.4
85      Roenis Elias  RP    55.333333      9.1
186   Kevin McCarthy  RP    53.666667     11.4
142  Jonathan Holder  RP    52.666667      8.0
248     Neil Ramirez  RP    47.333333      6.6
233      James Pazos  RP    45.666667      7.8
201      Juan Minaya  RP    42.666667      9.6
249     Neil Ramirez  RP    40.333333      6.6
310    Daniel Stumpf  RP    36.000000      9.4
231    David Paulino  RP    27.000000      6.0
119       Matt Grace  RP    23.000000      8.2
179      Matt Magill  RP    21.666667     11.1
189  Adalberto Mejia  RP     8.666667      9.1
225      Tyler Olson  RP     7.333333      6.6
128  Tayron Guerrero  RP     7.000000      7.3
229     Blake Parker  RP     2.000000      1.1

[159 rows x 4 columns]

## The Line-Up
- SP: Max Scherzer	
- RP: Chad Bettis

# Following One Player
Looking at one of our outfielders, **Nick Markakis** -  He earned a 'fantasy' score of 5534 and has a very low salary as compared to other cathcers. He had 623 AB with a BA of .297 for the 2018 season. He also hit 185 R, 43 doubles, 2 triples, and 14 HR during this season as well. This is the type of player who will allow us to rack up a lot of points for our fantasy league while still being a reasonably priced player.


## Limitations

Our main limitation to our approach will be our narrowed view on only assessing our position players based on salary, H, Doubles, Triples and HR. The same goes for our pitchers where we are only assessing on salary, IP, K, BB, HBP, and HR. There are other ways to earn points in Fantasy but we have decided these areas will give us the biggest bang for our buck. However, this may come to hurt us in the long run as we will most likely be under our budget. 